### PR TITLE
[ENH] Support partial term fetching failures

### LINKED ISSUE #54: Handle failed requests for available controlled terms gracefully
- Related to #38

Questions:
- should the response format be the same as for partial-success query results, e.g. with a section for the errors encountered?
- should "partial-success" terms fetching be allowed?

Decision:
If query tool fails to get terms from a node:
- For now, just return terms available from nodes that are reachable/don't error out
- Should keep the same response shape as for cohort queries e.g.:
```bash
{
  "nb:Assessment": [
    {
      "TermURL": "cogatlas:trm_56a9137d9dce1",
      "Label": "behavioral approach/inhibition systems"
    },
    {
      "TermURL": "cogatlas:trm_55a6a8e81b7f4",
      "Label": "Barratt Impulsiveness Scale"
    },
    ...
}
```
to:

```bash
{
  "errors": [{}] ...
  "responses":
  {
    "nb:Assessment": [
      {
        "TermURL": "cogatlas:trm_56a9137d9dce1",
        "Label": "behavioral approach/inhibition systems"
      },
      {
        "TermURL": "cogatlas:trm_55a6a8e81b7f4",
        "Label": "Barratt Impulsiveness Scale"
      },
      ...
  },
  "nodes_response_status": "..."
}
```

When request to all nodes fail:
```bash
{
  "errors": [<all errors>] ...
  "responses":
  {
    "nb:Assessment": []
  },
  "nodes_response_status": "fail"
}
```

### Steps to implement:
- [x] Switch to async requests to nodes
- [x] Add error responses to list
- [x] Add new response model for combined attribute response
- [ ] ~Rename tests of partial success _cohort_ queries to be more specific~
- [x] Consider refactoring mixed-status JSONResponse into its own class/function?
- [x] When requests to all nodes fail, should return `"responses": {<data_element_URI>: []}`

### PR DESCRIPTION
**Pull request by alyssadai**
- Closes #54

Changes proposed in this pull request:

- Add new response model for combined attribute response from all nodes
- Refactor out building of combined response into its own function
- Gracefully handle any errors from requested nodes 
- Switch to async process for sending out individual node requests for terms

## Checklist

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see https://neurobagel.org/contributing/pull_requests for more info)_
- [x] PR has a label for the release changelog or `skip-release` (to be applied by maintainers only)
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [x] Tests have been added

For bug fixes:
- [x] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

File: app/api/crud.py [hunk 1]
```diff
@@ -148,20 +156,36 @@ async def get_terms(data_element_URI: str):
...
+            node_errors.append(
+                {"node_name": node_name, "error": response.detail}
+            )
+            warnings.warn(
+                f"Request to node {node_name} ({node_url}) did not succeed: {response.detail}"
+            )
+        else:
+            # Build the dictionary of unique term-label pairings from all nodes
+            for term_dict in response[data_element_URI]:
+                unique_terms_dict[term_dict["TermURL"]] = term_dict
```
Comment (surchs) [inline app/api/crud.py:183]: If we overwrite previous instances here, would it be reasonable to check first if the term is the same as the one we've already stored? If we ever encounter a situation where some jokester has a neurobagel node where they just make up their own TermURLs for our terms, we could end up with a tricky to debug situation if all that determines what we return here is the order in which we unpack the responses from nodes. Wdyt?

Comment (surchs) [reply app/api/crud.py:183]: I just saw that this is old code, so let's not address this here. I'll make an issue to discuss in general

File: app/api/routers/attributes.py [hunk 2]
```diff
@@ -2,12 +2,12 @@
...
 
 from .. import crud
-from ..models import CONTROLLED_TERM_REGEX
+from ..models import CONTROLLED_TERM_REGEX, CombinedAttributeResponse
 
 router = APIRouter(prefix="/attributes", tags=["attributes"])
 
 
-@router.get("/{data_element_URI}")
+@router.get("/{data_element_URI}", response_model=CombinedAttributeResponse)
```
Comment (surchs) [inline app/api/routers/attributes.py:10]: in `crud`, you mentioned https://fastapi.tiangolo.com/advanced/additional-responses/ in a comment (very helpful). So I would expect that here you do have to differentiate between a full success response:

e.g.
```python
{
        "errors": node_errors,
        "responses": cross_node_results,
        "nodes_response_status": "success",
    }
```

and a partial or complete failure response, e.g.
```python
return JSONResponse(
            status_code=status.HTTP_207_MULTI_STATUS,
            content={
                "errors": node_errors,
                "responses": cross_node_results,
                "nodes_response_status": "partial success",
            },
        )
```

no? maybe I'm missing something about the additional responses or the different types of response objects here.

File: tests/conftest.py [hunk 3]
```diff
@@ -8,3 +9,11 @@
 def test_app():
     client = TestClient(app)
     yield client
+
+
+@pytest.fixture()
+def mock_failed_connection_httpx_get():
+    async def _mock_httpx_get_with_connect_error(self, **kwargs):
```
Comment (surchs) [inline tests/conftest.py:16]: :eyes:
a) like **kwargs here, I think we could use that with other fixtures where we're less concerned with the args
b) why `self` here?

File: tests/test_attributes.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+from app.api import utility as util
+
+
+def test_partially_failed_terms_fetching_handled_gracefully(
+    test_app, monkeypatch
+):
+    """
+    Test that when getting term instances for an attribute (/attribute/{data_element_URI}) fails for some nodes due to an error,
+    the overall API get request still succeeds, and the response includes a list of the encountered errors along with the successfully fetched terms.
```
Comment (surchs) [inline tests/test_attributes.py:13]:
```suggestion
    When some nodes fail while getting term instances for an attribute (/attribute/{data_element_URI}),
    the overall API get request still succeeds, and the response includes a list of the encountered errors along with the successfully fetched terms.
```

File: tests/test_attributes.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                "Label": "behavioral approach/inhibition systems",
+            },
+            {
+                "TermURL": "cogatlas:trm_55a6a8e81b7f4",
+                "Label": "Barratt Impulsiveness Scale",
+            },
+        ]
+    }
+
+    async def mock_httpx_get(self, **kwargs):
```
Comment (surchs) [inline tests/test_attributes.py:37]: not sure I understand `self` here, could you explain that?

File: tests/test_attributes.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    the overall API get request still succeeds, and the response includes a list of the encountered errors along with the successfully fetched terms.
+    """
+    monkeypatch.setattr(
+        util,
+        "FEDERATION_NODES",
+        {
+            "https://firstpublicnode.org/": "First Public Node",
+            "https://secondpublicnode.org/": "Second Public Node",
+        },
+    )
```
Comment (surchs) [inline tests/test_attributes.py:22]: I guess we'll need this for all the tests, right? Maybe we should lift all this into a fixture itself, see last example in https://docs.pytest.org/en/6.2.x/monkeypatch.html#monkeypatching-returned-objects-building-mock-classes

File: tests/test_attributes.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            kwargs["url"]
+            == "https://firstpublicnode.org/attributes/nb:Assessment"
+        ):
+            return httpx.Response(
+                status_code=200,
+                json=mocked_node_attribute_response,
+            )
+        return httpx.Response(
+            status_code=500, json={}, text="Some internal server error"
+        )
```
Comment (surchs) [inline tests/test_attributes.py:48]: :cherries: functionally the same, but I find it easier to read if the `if` statement just covers the fail / odd-one-out, and then the else path is the default / valid response

File: tests/test_attributes.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    assert response.json() == {
+        "errors": [
+            {
+                "node_name": "Second Public Node",
+                "error": "Internal Server Error: Some internal server error",
+            },
+        ],
+        "responses": mocked_node_attribute_response,
+        "nodes_response_status": "partial success",
+    }
```
Comment (surchs) [inline tests/test_attributes.py:66]: I would split these to deal with different concerns:

```suggestion
    response_object = response.json()
    assert response_object["errors"] == {
                "node_name": "Second Public Node",
                "error": "Internal Server Error: Some internal server error",
            }
    assert response_object["nodes_response_status"] == "partial success"
    assert response_object["responses"] == mocked_node_attribute_response
```
This makes tests more readable, also, I believe the test summary if only one of these should fail. 
And another reason to switch the if-else routes in the mock as mentioned above: it's a bit confusing to see ` "Second Public Node"` handled here when in the mock we only look at `https://firstpublicnode.org` - I suggest we make both align here by explicitly handling the failing node in the mock

File: tests/test_attributes.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def test_fully_failed_terms_fetching_handled_gracefully(
+    test_app, monkeypatch, mock_failed_connection_httpx_get
+):
+    """
+    Test that when getting term instances for an attribute (/attribute/{data_element_URI}) fails for *all* nodes,
+    the overall API get request still succeeds, but includes an overall failure status and all encountered errors in the response.
```
Comment (surchs) [inline tests/test_attributes.py:74]:
```suggestion
    When *all* nodes fail while getting term instances for an attribute (/attribute/{data_element_URI}),
    the overall API get request still succeeds, but includes an overall failure status and all encountered errors in the response.
```

**Review by surchs — COMMENTED** [review]
Very nice @alyssadai :tada:, thank you for dealing with this! I left a couple of comments  / suggestions. Let me know what you think

File: app/api/crud.py (hunk 1 above)
Comment (alyssadai) [reply app/api/crud.py:183]: Thanks for catching this. I couldn't tell if you already created this issue so here's a new one! https://github.com/neurobagel/federation-api/issues/68

Comment (surchs) [reply app/api/crud.py:183]: Indeed, I had not! Thanks for catching that :)

File: app/api/routers/attributes.py (hunk 2 above)
Comment (alyssadai) [reply app/api/routers/attributes.py:16]: See new commits for comments added to explain this!

File: tests/conftest.py (hunk 3 above)
Comment (alyssadai) [reply tests/conftest.py:32]: See new commits for comments added to explain this!

File: tests/test_attributes.py (hunk 5 above)
Comment (alyssadai) [reply tests/test_attributes.py:28]: See new commits for comments added to explain this!

**Review by surchs — APPROVED** [review]
very nice! thanks @alyssadai :tada: 

:cook:

**Comment by surchs** [comment]
:rocket: PR was released in [`v0.1.0`](https://github.com/neurobagel/federation-api/releases/tag/v0.1.0) :rocket:

**Pull request merged**